src/fe/fe.C

```diff
@@ -287,6 +287,13 @@ void FE<Dim,T>::reinit(const Elem * elem,
       {
         if (T != LAGRANGE)
           libmesh_warning("dual calculations have only been verified for the LAGRANGE family");
+
+        // In order for the matrices for the biorthogonality condition to be
+        // non-singular, the dual basis coefficients must be computed when the
+        // primal shape functions are evaluated with a quadrature rule
+        if (!pts)
+          this->compute_dual_shape_coeffs();
```

Comment:
Nice change here!

Reply:
That seems bad. The points being passed for reinitializing the lower-D element won't correspond to a quadrature rule for the lower-D element; they will correspond to a quadrature rule for the mortar element. I thought you told me that reinit needed to be called for the "full" element in order to compute the correct coefficients? We're going to end up back in the same boat where a user could very easily construct a singular matrix

Reply:
For the mortar approach where we integrate on the projected points, the non-singularity of the local matrices depends on the correctness of the projected integration points.. Please see https://github.com/idaholab/moose/pull/15216#discussion_r423129248

Reply:
So instead of `if (!pts)`, the check should be `if (you're giving me a set of points I approve of)` :smiley:

Reply:
Ugh, so I guess I'm going to have to remove this check and then just not use the preexisting `fe_test` for coverage :frowning_face:

Reply:
> So instead of `if (!pts)`, the check should be `if (you're giving me a set of points I approve of)` 😃

Haha that is definitely true! I am just not sure what is the mathematical equivalence of 'you're giving me a set of points I approve of' .. maybe putting a non-singularity check in the `compute_dual_coeff` would be the easiest fix

Reply:
Should we check (maybe only in debug mode or if the user requests it) whether the matrix that gets constructed is singular? Unfortunately DenseMatrix::det() overwrites the matrix with an LU decomposition to compute the determinant... so this check would not be cheap or fast.

Reply:
What do we do if the check fails? Warn the user and continue? Maybe the hard failure out of the LU factorization is the most appropriate thing

Reply:
Thanks John! We are actually doing `lu_solve` later in the code. So it is already doing a LU decomposition, can we make use of that? 

Or instead of computing the determinate, is there a warning in `lu_solve` so that we throw an error when the solve fails?

Reply:
> What do we do if the check fails? Warn the user and continue? Maybe the hard failure out of the LU factorization is the most appropriate thing

I agree. The code should throw an error and quit when the LU solve fails.

Reply:
OK, sorry I kind of forgot that you already do an LU solve, but yeah both the hand-coded and Lapack-based `DenseMatrix::_lu_decompose`, routines will throw an exception when the matrix is singular, so you could then potentially recover from that in some way at the MOOSE level.

Reply:
It may also be worth computing the `det()` after the `lu_solve()` is done since it's very cheap relative to the cost of `lu_solve()` (just the product of diagonal entries). This may alert you to cases which are nearly singular but which don't otherwise throw an exception... not sure how relevant that is in practice but I imagine it could affect the accuracy of the method if the `det()` is 1.e-12 or something...

Reply:
Oh, good point.  Using LU is currently required for that test+warning to work, because our DenseMatrix::det() implementation depends on LU, but it wouldn't be hard to just upgrade det().